engines/scumm/smush/smush_player.cpp

```diff
@@ -1304,4 +1279,770 @@ void SmushPlayer::play(const char *filename, int32 speed, int32 offset, int32 st
+	}
+
+	for (int i = 0; i < _smushNumTracks; i++) {
+		free(_smushTracks[i].blockPtr);
+		free(_smushTracks[i].fadeBuf);
+	}
+	_smushNumTracks = 0;
+}
+
+int SmushPlayer::isChanActive(int flagId) {
+	return _smushTrackFlags[flagId];
+}
+
+int SmushPlayer::setChanFlag(int id, int flagVal) {
+	if (id >= 0 && id <= _smushNumTracks)
+		_smushTrackFlags[id] = flagVal;
+
+	return id;
+}
+
+int SmushPlayer::addAudioTrack(int32 trackBlockSize, int32 maxBlockSize) {
+	int id = _smushNumTracks;
+	_smushTracks[id].state = TRK_STATE_INACTIVE;
+	_smushTracks[id].audioRemaining = 0;
+	_smushTracks[id].flags = 0;
+	_smushTracks[id].groupId = GRP_MASTER;
+	_smushTracks[id].blockSize = trackBlockSize;
+	_smushTracks[id].parsedChunks = 0;
+
+	_smushTracks[id].fadeBuf = (uint8 *)malloc(SMUSH_FADE_SIZE);
+	if (!_smushTracks[id].fadeBuf)
+		return -1;
+
+	_smushTracks[id].blockPtr = (uint8 *)malloc(_smushTracks[id].blockSize);
+	if (!_smushTracks[id].blockPtr)
+		return -1;
+
+	fillAudioBufWithSample(_smushTracks[id].blockPtr, 0, _smushTracks[id].blockSize);
+	// Track the effective number of tracks, so that if only one fails,
+	// the others can carry on as they should
+	_smushNumTracks++;
+
+	return 0;
+}
+
+void SmushPlayer::fillAudioBufWithSample(uint8 *buf, int8 sampleVal, uint32 size) {
+	memset(buf, sampleVal, size);
+	memset(&buf[4 * (size >> 2)], sampleVal, size & 3);
+}
+
+void SmushPlayer::resetAudioTracks() {
+	for (int i = 0; i < _smushNumTracks; i++) {
+		_smushTracks[i].state = TRK_STATE_INACTIVE;
+		_smushTracks[i].groupId = GRP_MASTER;
+	}
+}
+
+void SmushPlayer::setGainReductionParams(int16 reductionThreshold, int16 reductionConst) {
+	if (reductionThreshold)
+		_reductionThresh = reductionThreshold;
+	if (reductionConst)
+		_gainReductionConst = reductionConst;
+}
+
+void SmushPlayer::setGroupVolume(int groupId, int volValue) {
+	switch (groupId) {
+	case GRP_MASTER:
+		_smushTrackVols[0] = volValue; // Master
+		break;
+	case GRP_SFX:
+		_smushTrackVols[1] = volValue; // Sfx
+		break;
+	case GRP_BKGMUS:
+		_smushTrackVols[3] = volValue; // Music
+		break;
+	case GRP_SPEECH:
+		_smushTrackVols[2] = volValue; // Voice
+		break;
+	default: // Internal groups
+		for (int i = 0; i < SMUSH_MAX_TRACKS; i++) {
+			if (_smushTracks[i].groupId == groupId)
+				_smushTracks[i].volume = volValue;
+		}
+		break;
+	}
+}
+
+void SmushPlayer::handleSAUDChunk(uint8 *srcBuf, uint32 size, int groupId, int vol, int pan, int16 flags, int trkId, int index, int maxFrames) {
+	int targetTrk, diff1, diff2;
+	uint32 dataSize, mod;
+	int32 maxBlockSize, saudSize;
+	uint16 flagsAccumulator[SMUSH_MAX_TRACKS];
+
+	if (!index) {
+		saudSize = READ_BE_UINT32(&srcBuf[4]);
+
+		// Find the maximum block sizes between the current ones
+		maxBlockSize = 0;
+		for (int i = 0; i < _smushNumTracks; i++) {
+			if (_smushTracks[i].blockSize > maxBlockSize) {
+				maxBlockSize = _smushTracks[i].blockSize;
+			}
+		}
+
+		diff2 = 10000000;
+		if ((flags & 0xC0) != IS_BKG_MUSIC) {
```

Comment:
if you using macro, so I guess both should have defined macro

Reply:
Okay, so you mean something like `(flags & TRK_TYPE_FLAG)`?

Reply:
Okay, I will do that, thanks